### ISSUE TITLE
Error: SecurityError: DOM Exception 18

### ISSUE THREAD

**Issue by kevinchiu** [tsayen/dom-to-image#40]
> Error: SecurityError: DOM Exception 18

I seem to be getting this error only in Safari when I try to save a node. There is an image in the node, but it is from the same domain as the website so this should not be a CORS issue.

**Comment by teamco** [comment]
The same issue
Error: SecurityError: DOM Exception 18
code: 18
column: 648
line: 30
message: "SecurityError: DOM Exception 18"
name: "SecurityError"
stack: "toDataURL@[native code]↵↵promiseReactionJob@[native code]"

**Comment by svenvandescheur** [comment]
I also ran into this issue, has anyone managed to work around it yet?

**Comment by tsayen** [comment]
could you please put together a jsfiddle that fails for you?

**Comment by acoyfellow** [comment]
Still digging, but this fiddle will re-produce "SecurityError: DOM Exception 18" for me, on desktop Safari, mobile everything (Opera, FF, Chrome, Safari), but it _works_ as expected with Chrome on desktop.

https://jsfiddle.net/acoyfellow/Lggtwjbb/1/

**Comment by tsayen** [comment]
sorry to say, but Icannot reproduce that since don't own a Mac :smile:  
anyone with a mac want to make a PR fixing this issue?

**Comment by thienhv295** [comment]
The same issue :(. Has anyone fix it?

**Comment by kevinchiu** [comment]
@tsayen You can try using the Midori browser on Linux or Windows: http://midori-browser.org/

**Comment by WingGao** [comment]
After searching the web, I couldn't find a satisfactory solution.
So I just use a SVG to replace.
```js
//line95
    function toPng(node, options) {
        return draw(node, options || {})
            .then(function (result) {
                try {
                    var image = result.canvas.toDataURL("image/png");
                    return image;
                } catch (err) {
                    console.warn(err);
                    return result.svg.src;
                }
            });
    }
```
```js
//line 135
function draw(domNode, options) {
        return toSvg(domNode, options)
            .then(util.makeImage)
            .then(util.delay(100))
            .then(function (image) {
                var canvas = newCanvas(domNode);
                // after this, Canvas’ origin-clean is DIRTY
                // safari will throw 'SecurityError: DOM Exception 18'
                // but it can use 'data:images/svg+xml'
                canvas.getContext('2d').drawImage(image, 0, 0);
                return {canvas: canvas, svg: image};
            });
...
```

**Comment by thienhv295** [comment]
@WingGao: Actually, I had changed like you. And then it's show on browser only, but I cannot save image to file on server (cannot save data:images/svg+xml to image file). Do you have any solution for this issue?

**Comment by WingGao** [comment]
@thienhv295 Unfortunately, I used the server to decode the SVG but failed.
Finally, I change the lib to html2canvas...It seems work to me...

**Comment by EvianZhow** [comment]
@WingGao I just wonder why html2canvas works for you. I'm using that lib but encountered the same error: `SecurityError: DOM Exception 18`. Can you provide a caveat about your code?

**Comment by WingGao** [comment]
@EvianZhow 
OK, maybe you miss the option `useCORS`.
```js
html2canvas(myElement, {useCORS: true}).then(function (canvas) {
                            try {
                                b64Data = canvas.toDataURL("image/png");
                            } catch (err) {
                                console.log(err)
                            }
                        }).catch(function onRejected(error) {
                        });
```

**Comment by EvianZhow** [comment]
@WingGao My caveat is similar to your's. But Safari raise the exception: `SecurityError (DOM Exception 18): The operation is insecure.`

**Comment by mtmckenna** [comment]
Hello! I'm getting Exception 18 in Safari too. Here's the error:

```
Error: SecurityError: DOM Exception 18
toDataURL
(anonymous function) — dom-to-image.js:98
promiseReactionJob
```

Might be related to this issue? http://stackoverflow.com/questions/33972254/svgpng-from-canvas-todataurl-throws-dom-exception-18-security-error-in-safari-9

**Comment by WingGao** [comment]
@EvianZhow @mtmckenna 
Here is my demo, I hope it can help you.
https://jsfiddle.net/hkkynoyz/

**Comment by MarkusSvensson** [comment]
I have this problem and it is blocking my development for as it is an Ionic app for iOS... 

I verified @WingGao's "solution" to use html2canvas - it works but have some rendering problems and html2canvas does not seem to be maintained anymore.

Just want to contribute to prioritisation for this issue. As it works for html2canvas there 'should' be a workaround, but I don't know where to start.

**Comment by Digity** [comment]
I tried something that worked based on this post:
http://stackoverflow.com/questions/2390232/why-does-canvas-todataurl-throw-a-security-exception

In dom-to-image.js there is the following code, but I added the line  `image.setAttribute('crossOrigin','anonymous');`

It worked and I didn't get the expection anymore.  I guess the point is that maybe this library should provide options for setting the 'crossOrigin' property on the image.

```
 function makeImage(uri) {
            return new Promise(function (resolve, reject) {
                var image = new Image();
                image.setAttribute('crossOrigin','anonymous');
                image.onload = function () {
                    resolve(image);
                };
                image.onerror = reject;
                image.src = uri;
            });
}
```

**Comment by Digity** [comment]
I'm not sure the dom-to-image library can be changed to fix this issue.  Isn't the issue more with the canvas being "tainted" security wise and therefore, it'll throw the exception?  I think the solution is to find out how the canvas is being tainted and then fix it so it no longer violates the security rules.  How it's tainted could vary from one example to the other.  Am I wrong about this assessment?

**Comment by Digity** [comment]
I'm having the same problem and I hope there's a solution soon.

**Comment by pietersv** [comment]
The Safari security exception in `toDataURL()` is thrown in Safari because the image URL doesn't match the page domain.  I don't know why it's not thrown in the other browsers.  However, the call to `canvas.toDataURL` is unnecessary, we can work around it as follows:

First modify the function `toSvg()` to append the `image` string directly to `canvas` on line 133 as follows:
```js
function draw(domNode, options) {
    return toSvg(domNode, options)
        .then(util.makeImage)
        .then(util.delay(100))
        .then(function (image) {
          var canvas = newCanvas(domNode);
          canvas.getContext('2d').drawImage(image, 0, 0);
          canvas.image = image;  // attach image string so we can recover it without .toDataURL()
          return canvas;
        });
   ...
  });
```

then on line 98 retrieve this string directly
```js
  function toPng(node, options) {
    return draw(node, options || {})
        .then(function (canvas) {
          return canvas.image.src;  // was canvas.toDataURL();
        });
  }
```

**Comment by Digity** [comment]
@pietersv I don't think that works, because it's not converting it to png.  The "toDataUrl" does the png conversion.  You're skipping the png conversion step.

**Comment by Digity** [comment]
I came up with a bit of a workaround.  Basically, I use the toSvg method and then I take the <svg>....</svg> part and wrap it in <html><body>...</body></html>.  I then pass this to webshot library (https://github.com/brenden/node-webshot), which uses phantomjs to take the html and create a screenshot of it.  It works, but obviously not the ideal solution.

**Comment by pietersv** [comment]
@digity You're right, this worked for part of the ways I used `toDataURL()` but not the actual PNG file. Wishful thinking.

**Comment by Digity** [comment]
Here's some quick sample code for converting it to png using webshot library:
```
domtoimage.toSvg(node)
.then(function(dataUrl) {
	var svg = dataUrl.substring(dataUrl.indexOf('<svg'));
	webshot('<html><body><div>'+svg+'</div></body></html>', 'myimage.png', {siteType:'html'}, function(err) {
		// screenshot now saved to myimage.png
	});
})
.catch(function(error) {
	...
});
```

**Comment by kapyaar** [comment]
I am facing  the same issue. Even without any images, just pure text, it is throwing a Dom exception 18 error. 

http://jsfiddle.net/8h13tmcj/26/

**Comment by kapyaar** [comment]
@tsayen I think it is the <foreignObject> tag causing the problem. Is it possible to generate the svg without this tag?

**Comment by laflower** [comment]
Same issue on ionic

**Comment by dashokkumar93** [comment]
If you use foreign object in the svg and then converting it to canvas which contains foreign object will cause security error.

**Comment by denis-sokolov** [comment]
Also see #74.

**Comment by denis-sokolov** [comment]
This is a duplicate of #27.

**Comment by Rainpia** [comment]
@WingGao I use this like you . but can not save image, do you have some advise ?

**Comment by alanhe421** [comment]
@Rainpia @WingGao how to save pic ?

**Comment by bsautel** [comment]
@alanhg , see [this comment](https://github.com/tsayen/dom-to-image/issues/74#issuecomment-270863365) in which I explain the workaround I use to make a screen shot in a Cordova application. It supports Android and iOS only but it may be extended to support additionnal browsers.

Hope it will help you!

**Comment by minchangyong** [comment]
who can copy once demo in there！
